server/src-lib/Hasura/Server/Auth/Utils.hs

```diff
@@ -0,0 +1,52 @@
+module Hasura.Server.Auth.Utils
+  ( timeFromCacheControlHeader
+  , timeFromExpiresHeader
+  , toUTCTime
+  , toNominalDiffTime
+  ) where
+
+import           Data.Parser.CacheControl (parseMaxAge)
+import           Data.Time.Clock
+import           Data.Time.Format         (defaultTimeLocale, parseTimeM)
+
+import qualified Data.String.Conversions  as CS
+import qualified Data.Text                as T
+
+import           Hasura.Prelude
+
+
+data ExpiryTime = AbsoluteTime UTCTime
+                | RelativeTime NominalDiffTime
+
+
+-- | Extracts a relative expiry time from the CacheControl header.
+timeFromCacheControlHeader
+  :: (Monad m)
+  => Maybe T.Text         -- ^ the header
+  -> (T.Text -> m ())     -- ^ what should be done if parsing fails
```

Comment:
This seems like a curious way to model this to me! Why not give these functions the following simpler type signature?

```haskell
timeFromCacheControlHeader, timeFromExpiresHeader :: MonadError Text m => Text -> m ExpiryTime
```

Then the caller can figure out what to do in the case of a `Maybe`, since neither of these functions do anything interesting when `Nothing` is passed, anyway, and the caller can instantiate `m` to `Either Text` if they want to do something custom on errors.

Reply:
Done!